internal/cloud/backend_apply.go

```diff
@@ -133,6 +133,19 @@ func (b *Cloud) opApply(stopCtx, cancelCtx context.Context, op *backend.Operatio
 		}
 	}
 
+	// Retrieve the run to get task stages.
+	// Task Stages are calculated upfront so we only need to call this once for the run.
+	taskStages, err := b.runTaskStages(stopCtx, b.client, r.ID)
+	if err != nil {
+		return r, err
+	}
```

Comment:
One thing I'm minorly concerned about is just how many times we actually read a run and how we read the run _again_ to fetch the task stages within each backend operation. If you notice in L78 we return the run from the plan operation, then we fetch the run again in L91. And you fetch the run one more time including the task stages here in L138. If you look into the plan operation, we fetch the run multiple times:

https://github.com/glennsarti/terraform/blob/1773129832894a90bbc90f93759b8494981baced/internal/cloud/backend_plan.go#L296-L305

and here

https://github.com/glennsarti/terraform/blob/1773129832894a90bbc90f93759b8494981baced/internal/cloud/backend_plan.go#L339-L343

(This is obviously excluding the times we fetch the run's status). 

Effectively, since task stages are immutable after a run is created (?) should we only fetch them once? That is after the run is created and reuse the fetched stages across operations? It might not be a trivial refactor. What do you think?

Reply:
https://github.com/glennsarti/terraform/blob/1773129832894a90bbc90f93759b8494981baced/internal/cloud/backend_plan.go#L339-L343 is not used for the task stage logic, that's for cost estimation and policy checks.

> Effectively, since task stages are immutable after a run is created (?) should we only fetch them once?
Yeah, that is true.

Although all of the other run queries don't include the task stages in the response.

>  It might not be a trivial refactor. What do you think?
I think it's not trivial right now. We'd need a way to feed in the immutable run information into the backend_apply and backend_plan (A new context perhaps? or a function arg... dunno which is best. It's not crossing go routines AFAIK).

And we'd need to split out the run creation logic from backend_plan into something else (backend_create_run).

That's alot of work **right now** to save up to 2 (relatively small) API calls.

Reply:
Yeah that's fair, a bit of a slight overreaction on my part. I figured since we read the run multiple times anyways, we could replace any of those API calls with `ReadWithOptions()`, include the stages, and store them in a local variable without having to make an extra API call when we need to fetch the stage IDs. 

I'll smoke test tomorrow. Thanks for the thoughtful response!

Reply:
Hi @sebasslash  just checking in, how'd the smoke test go?